### ISSUE TITLE
Eqx call cannot return tensoflow-probability (TFP) distribution 

### ISSUE THREAD

**Issue by peabody124** [patrick-kidger/equinox#911]
I have a project that is using TFP distributions fairly extensively. It works when I only eqx.filter_jit over the loss equation but if I start trying to wrapper some components I get errors. 

I'm guessing this is expected behavior since the TFP distributions aren't appropriate pytree's but I wanted to confirm and see if there is any type of workaround. I'm guessing https://github.com/lockwo/distreqx is the best plan, although it doesn't have some of the distributions like a low rank diagnoal.

Simple unit test to replicate this where the eqx.filter_jit fails:

```
import jax
import equinox as eqx
from jax import numpy as jnp
import tensorflow_probability.substrates.jax as tfp

tfd = tfp.distributions
tfb = tfp.bijectors


class SimpleTFD(eqx.Module):
    """
    Wraps an ImplicitTrajectory to produce a multivariate normal distribution
    using a diagonal plus low rank covariance matrix.
    """

    mu: jnp.array
    d: jnp.array

    def __init__(
        self,
        mu: jnp.array,
        d: jnp.array,
    ):

        self.mu = mu
        self.d = d

    def __call__(
        self,
        o: jnp.ndarray,
    ) -> jnp.ndarray:

        mvn = tfd.MultivariateNormalDiag(loc=self.mu + o, scale_diag=self.d, validate_args=True, allow_nan_stats=False)

        # # return this as a pytree by converting the fields in the dataclass to a dictionary
        # fields = self.__dataclass_fields__
        # mvn = {k: getattr(mvn, k) for k in fields}

        return mvn


class SimpleTFD2(eqx.Module):
    """
    Wraps an ImplicitTrajectory to produce a multivariate normal distribution
    using a diagonal plus low rank covariance matrix.
    """

    mu: jnp.array
    d: jnp.array

    def __init__(
        self,
        mu: jnp.array,
        d: jnp.array,
    ):

        self.mu = mu
        self.d = d

    def __call__(
        self,
        o: jnp.ndarray,
    ) -> jnp.ndarray:

        mvn = tfd.MultivariateNormalDiag(loc=self.mu + o, scale_diag=self.d, validate_args=True, allow_nan_stats=False)

        return mvn.mean()


def test_construction():
    mu = jnp.array([1.0, 2.0])
    d = jnp.array([1.0, 2.0])
    model = SimpleTFD(mu, d)
    assert jnp.all(model.mu == mu)
    assert jnp.all(model.d == d)


def test_call():
    mu = jnp.array([1.0, 2.0])
    d = jnp.array([1.0, 2.0])
    model = SimpleTFD(mu, d)
    o = jnp.array([1.0, 2.0])
    res = model(o)

    assert isinstance(res, tfd.MultivariateNormalDiag)


def test_jit_call():
    ################### **THIS FAILS**
    mu = jnp.array([1.0, 2.0])
    d = jnp.array([1.0, 2.0])
    model = SimpleTFD(mu, d)
    o = jnp.array([1.0, 2.0])
    res = eqx.filter_jit(model)(o)

    assert isinstance(res, tfd.MultivariateNormalDiag)


def test_jit_mean_call():
    mu = jnp.array([1.0, 2.0])
    d = jnp.array([1.0, 2.0])
    model = SimpleTFD2(mu, d)
    o = jnp.array([1.0, 2.0])
    res = eqx.filter_jit(model)(o)
```

Thanks for all the work on equinox. it is great!

**Comment by patrick-kidger** [comment]
Yup, if TFP distributions aren't pytrees then there's little we can do.

I'd definitely recommend distreqx as a good choice, and I'll tag @lockwo here if there's anything missing from it that you'd find useful.

**Comment by peabody124** [comment]
Cool, thanks for clarifying! 

In case it is helpful for anyone, here is a lightweight wrapper I'm using until I migrate to distreqx

```
import abc
import equinox as eqx
from jax import numpy as jnp

import tensorflow_probability.substrates.jax as tfp

tfd = tfp.distributions
tfb = tfp.bijectors


def ProbabilityWrapper(cls):

    class WrappedClass(cls):

        def __init__(self, *args, **kwargs):
            super().__init__(*args, **kwargs)

        def mean(self):
            return self.dist.mean()

        def entropy(self):
            return self.dist.entropy()

        def sample(self, *args, **kwargs):
            return self.dist.sample(*args, **kwargs)

        def log_prob(self, *args, **kwargs):
            return self.dist.log_prob(*args, **kwargs)

        def covariance(self):
            return self.dist.covariance()

        def stddev(self):
            return self.dist.stddev()

        @property
        def batch_shape(self):
            return self.dist.batch_shape

    return WrappedClass


@ProbabilityWrapper
class Normal(eqx.Module):

    loc: jnp.array
    scale: jnp.array

    def __init__(
        self,
        loc: jnp.array,
        scale: jnp.array,
    ):

        self.loc = loc
        self.scale = scale

    @property
    def dist(self):
        return tfd.Normal(loc=self.loc, scale=self.scale)


@ProbabilityWrapper
class Laplace(eqx.Module):

    loc: jnp.array
    scale: jnp.array

    def __init__(
        self,
        loc: jnp.array,
        scale: jnp.array,
    ):

        self.loc = loc
        self.scale = scale

    @property
    def dist(self):
        return tfd.Laplace(loc=self.loc, scale=self.scale)


@ProbabilityWrapper
class MultivariateNormalDiag(eqx.Module):

    loc: jnp.array
    scale_diag: jnp.array

    def __init__(
        self,
        loc: jnp.array,
        scale_diag: jnp.array,
    ):

        self.loc = loc
        self.scale_diag = scale_diag

    @property
    def dist(self):
        return tfd.MultivariateNormalDiag(loc=self.loc, scale_diag=self.scale_diag, validate_args=True, allow_nan_stats=False)


@ProbabilityWrapper
class MultivariateNormalDiagPlusLowRank(eqx.Module):

    loc: jnp.array
    scale_diag: jnp.array
    scale_perturb_factor: jnp.array

    def __init__(
        self,
        loc: jnp.array,
        scale_diag: jnp.array,
        scale_perturb_factor: jnp.array,
    ):

        self.loc = loc
        self.scale_diag = scale_diag
        self.scale_perturb_factor = scale_perturb_factor

    @property
    def dist(self):
        return tfd.MultivariateNormalDiagPlusLowRank(
            loc=self.loc,
            scale_diag=self.scale_diag,
            scale_perturb_factor=self.scale_perturb_factor,
            scale_perturb_diag=None,
            validate_args=False,
            allow_nan_stats=False,
        )

```

which ends up passing the above unit tests

**Comment by lockwo** [comment]
Hey, that is definitely the goal of https://github.com/lockwo/distreqx, although as you noted its in early stages. There's still quite some work before its the fully fledged replacement of distrax I would like (and it's mostly a hobby project so I don't have a ton of time to add all I would want to add), but if you have specific distributions/bijectors that you want sooner rather than later I'm happy to accept PRs and/or accept issues telling me what you want and I'll just make them the distributions I do next (I am traveling over the holidays, so there will be quite some lag time tho).
